domains/pallets/auto-id/src/lib.rs

```diff
@@ -0,0 +1,519 @@
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Issuer auto id does not exist.
+        UnknownIssuer,
+        /// Issuer is not a root certificate.
+        IssuerNotRoot,
+        /// Certificate is invalid,
+        InvalidCertificate,
+        /// Invalid certificate validity.
+        InvalidValidity(ValidityError),
+        /// Invalid signature.
+        InvalidSignature,
+        /// AutoId identifier overflow.
+        IdentifierOverflow,
+        /// Certificate serial already issued.
+        CertificateSerialAlreadyIssued,
+        /// Certificate expired.
+        ExpiredCertificate,
+        /// Certificate revoked.
+        CertificateRevoked,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub (super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Emits when a new AutoId is registered.
+        NewAutoIdRegistered(Identifier),
+        /// Emits when a Certificate associated with AutoId is revoked.
+        CertificateRevoked(Identifier),
+        /// Emits when an AutoId is deactivated.
+        AutoIdDeactivated(Identifier),
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Registers a new AutoId after validating the provided certificate.
+        #[pallet::call_index(0)]
+        // TODO: benchmark
+        #[pallet::weight({10_000})]
+        pub fn register_auto_id(origin: OriginFor<T>, req: RegisterAutoId) -> DispatchResult {
+            ensure_signed(origin)?;
+            Self::do_register_auto_id(req)?;
+            Ok(())
+        }
+
+        /// Revokes a certificate associated with given AutoId.
+        #[pallet::call_index(1)]
+        // TODO: benchmark
+        #[pallet::weight({10_000})]
+        pub fn revoke_certificate(
+            origin: OriginFor<T>,
+            auto_id_identifier: Identifier,
+            signature: Signature,
+        ) -> DispatchResult {
+            ensure_signed(origin)?;
+            Self::do_revoke_certificate(auto_id_identifier, signature)?;
+            Ok(())
+        }
+
+        /// Deactivates a given AutoId.
+        #[pallet::call_index(2)]
+        // TODO: benchmark
+        #[pallet::weight({10_000})]
+        pub fn deactivate_auto_id(
+            origin: OriginFor<T>,
+            auto_id_identifier: Identifier,
+            signature: Signature,
+        ) -> DispatchResult {
+            ensure_signed(origin)?;
+            Self::do_deactivate_auto_id(auto_id_identifier, signature)?;
+            Ok(())
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    pub(crate) fn do_register_auto_id(req: RegisterAutoId) -> DispatchResult {
+        let current_time = T::Time::now();
+        let certificate = match req {
+            RegisterAutoId::X509(x509_req) => match x509_req {
+                RegisterAutoIdX509::Root {
+                    certificate,
+                    signature_algorithm,
+                    signature,
```

Comment:
Can this data be included in `certificate` rather than sent separately?

Reply:
The actual X509 certificate structure is as follows
```
X509Certificate{
	tbs_certificate
	signature_algorithm
	signature
}
```

The structure you are seeing there is deconstructed one so that runtime does not need to decode the full certificate and encode it back again and we dont need to hold the signature details once the verification is done. We only need to hold the tbs_certificate which contain actual X509 cert details